### DISCUSSION TITLE
Can I somehow share memory between dependency modules?

### DISCUSSION THREAD

**Discussion by NinevskiyK** [bytecodealliance/wasm-micro-runtime#2363] [Q&A] [answered]
I've gone through [Multiple Modules As Dependencies](https://wamr.gitbook.io/document/wamr-in-practice/features/multi_module) page and tried to make my own dependencies, but found that I can't simply pass buffer from a main module to a submodule. 

In particular, I have this two piece of code:
```
// main.c
__attribute__((import_module("worker")))
__attribute__((import_name("worker"))) extern int
worker(void *mas);

int compute_result(char *mas) {
    return worker(mas);
}

int main () {}
```
and
```
// worker.c
__attribute__((export_name("worker"))) int
worker(char *mas) {
    return mas[0];
}
```

And it seems that worker only sees zeros instead of the actual data. I've created the buffer with `wasm_runtime_module_malloc`. 
Why is that? How can I deal with that?

**Comment by lum1n0us** [comment, marked as answer]
Because every instance has its own liner memory in this case, *worker* can not see the content in *main*'s memory directly.

Using `wasm_runtime_module_malloc` is one possible way. It needs additional help from host. In that scenario, host will manage the lifecyle of `mas` (allocation and release). Host will dispatch `mas` to `worker` and other processores(if exist).

**Reply by NinevskiyK** [reply]
> `wasm_runtime_module_malloc` is one possible way

Am I right that the provided solution is to allocate another buffer(`mas1`) with the `wasm_runtime_module_malloc ` in the `worker` module on the host and then copy `mas` to `mas1`(on the host too)?

**Reply by lum1n0us** [reply]
It depends on the worker's responsibility.

Anyway, allocate another buffer(`mas1`) with the `wasm_runtime_module_malloc` or Wasm exported `malloc` in the host. It makes sure that the buffer is in the linear memory. And then, either let the host fill the content of the buffer and *worker* can read the content or pass the empty buffer to *worker* and *worker* will write the content into the buffer.

**Reply by NinevskiyK** [reply]
Yeah, I see. Thanks for answering my question!
